conda/notices.py

```diff
@@ -0,0 +1,502 @@
+        elif level == NoticeLevel.INFO:
+            return cls.OKGREEN
+        elif level == NoticeLevel.WARNING:
+            return cls.WARNING
+        else:
+            return cls.OKBLUE
+
+    @classmethod
+    def wrap_style(cls, text: Any, style: "TerminalStyle") -> str:
+        return f"{style}{text}{TerminalStyle.ENDC}"
+
+
+def print_notices(channel_notices: Sequence[ChannelNotice], ansi_colors: bool = True):
+    """
+    Accepts a list of channel notice responses and prints a display.
+    """
+    cur_chn = None
+
+    for ntc in channel_notices:
+        if cur_chn != ntc.channel_name:
+            print()
+            if ansi_colors:
+                channel_header = TerminalStyle.wrap_style("Channel:", TerminalStyle.BOLD)
+            else:
+                channel_header = "Channel:"
+            channel_header += f" {ntc.channel_name}"
+            print(channel_header)
+            cur_chn = ntc.channel_name
+        print_notice_message(ntc, ansi_colors=ansi_colors)
+        print()
+
+
+def print_notice_message(
+    notice: ChannelNotice, indent: str = "  ", ansi_colors: bool = True
+) -> None:
+    """
+    Prints a single channel notice
+    """
+    timestamp = "" if not notice.created_at else get_locale_timestamp(notice.created_at)
+
+    if ansi_colors:
+        level_style = TerminalStyle.get_color_from_level(notice.level)
+        level_text = TerminalStyle.wrap_style(notice.level, level_style)
+        level = f"[{level_text}]"
+
+        if notice.created_at:
+            level += TerminalStyle.wrap_style(f" -- {timestamp}", TerminalStyle.GRAY)
+    else:
+        level = f"[{notice.level}] -- {timestamp}"
+
+    print(f"{indent}{level}\n{indent}{notice.message}")
+
+
+def print_more_notices_message(
+    total_notices: int, displayed_notices: int, viewed_notices: int
+) -> None:
+    """
+    Conditionally shows a message informing users how many more message there are.
+    """
+    notices_not_shown = total_notices - viewed_notices - displayed_notices
+
+    if notices_not_shown > 0:
+        if notices_not_shown > 1:
+            msg = f"There are {notices_not_shown} more messages. " "To retrieve them run:\n\n"
+        else:
+            msg = f"There is {notices_not_shown} more message. " "To retrieve it run:\n\n"
+        print(f"{msg}conda notices\n")
+
+
+def get_locale_timestamp(timestamp: datetime) -> str:
+    """
+    Returns best attempt at a locale aware timestamp.
+    """
+    return timestamp.strftime("%c")
+
+
+def get_channel_name_and_urls(channels: [Sequence[Channel]]) -> Sequence[Tuple[str, str]]:
+    """Return a sequence of Ch"""
+
+    def ensure_endswith(value: str, ends: str) -> str:
+        return value if value.endswith(ends) else f"{value}{ends}"
+
+    def join_url(value: str, join_val: str) -> str:
+        return parse.urljoin(ensure_endswith(value, "/"), join_val)
+
+    return tuple(
+        (join_url(chn_url, NOTICES_FN), chn_obj.name or chn_obj.location)
```

Comment:
`urllib.parse.urljoin` tolerates duplicate trailing `/` and will deduplicate them for us:

```python
parse.urljoin("foo//", "bar")
'foo/bar'
```

```suggestion
    return tuple(
        (parse.urljoin(f"{chn_url}/", NOTICES_FN), chn_obj.name or chn_obj.location)
```

Reply:
@kenodegard, this has been left in there for the case of file URLs.

Example:
```python
>>> parse.urljoin('file:///Users/thath/', 'hello.php')  # behavior we want with a trailing slash
'file:///home/thath/hello.php'
>>> parse.urljoin('file:///Users/thath', 'hello.php')  # behavior we don't want without a trailing slash
'file:///home/hello.php'
```